processing/src/main/java/org/apache/carbondata/processing/newflow/steps/DataWriterBatchProcessorStepImpl.java

```diff
@@ -0,0 +1,206 @@
+      throw new CarbonDataLoadingException(
+          "Error while initializing data handler : " + e.getMessage());
+    } catch (Exception e) {
+      LOGGER.error(e, "Failed for table: " + tableName + " in DataWriterProcessorStepImpl");
+      throw new CarbonDataLoadingException("There is an unexpected error: " + e.getMessage());
+    }
+    return null;
+  }
+
+  @Override protected String getStepName() {
+    return "Data Writer";
+  }
+
+  private void finish(String tableName, CarbonFactHandler dataHandler) {
+    try {
+      dataHandler.finish();
+    } catch (Exception e) {
+      LOGGER.error(e, "Failed for table: " + tableName + " in  finishing data handler");
+    }
+    CarbonTimeStatisticsFactory.getLoadStatisticsInstance().recordTotalRecords(rowCounter.get());
+    processingComplete(dataHandler);
+    CarbonTimeStatisticsFactory.getLoadStatisticsInstance()
+        .recordDictionaryValue2MdkAdd2FileTime(configuration.getPartitionId(),
+            System.currentTimeMillis());
+    CarbonTimeStatisticsFactory.getLoadStatisticsInstance()
+        .recordMdkGenerateTotalTime(configuration.getPartitionId(), System.currentTimeMillis());
+  }
+
+  private void processingComplete(CarbonFactHandler dataHandler) throws CarbonDataLoadingException {
+    if (null != dataHandler) {
+      try {
+        dataHandler.closeHandler();
+      } catch (CarbonDataWriterException e) {
+        LOGGER.error(e, e.getMessage());
+        throw new CarbonDataLoadingException(e.getMessage());
+      } catch (Exception e) {
+        LOGGER.error(e, e.getMessage());
+        throw new CarbonDataLoadingException("There is an unexpected error: " + e.getMessage());
+      }
+    }
+  }
+
+  private void processBatch(CarbonRowBatch batch, CarbonFactHandler dataHandler,
+      SegmentProperties segmentProperties)
+      throws CarbonDataLoadingException {
+    int batchSize = 0;
+    try {
+      while (batch.hasNext()) {
+        CarbonRow row = batch.next();
+        batchSize++;
+        Object[] outputRow;
+        // adding one for the high cardinality dims byte array.
+        if (noDictionaryCount > 0 || complexDimensionCount > 0) {
+          outputRow = new Object[measureCount + 1 + 1];
+        } else {
+          outputRow = new Object[measureCount + 1];
+        }
+
+        int l = 0;
+        int index = 0;
+        Object[] measures = row.getObjectArray(measureIndex);
+        for (int i = 0; i < measureCount; i++) {
+          outputRow[l++] = measures[index++];
+        }
+        outputRow[l] = row.getObject(noDimByteArrayIndex);
+
+        int[] highCardExcludedRows = new int[segmentProperties.getDimColumnsCardinality().length];
+        int[] dimsArray = row.getIntArray(dimsArrayIndex);
+        for (int i = 0; i < highCardExcludedRows.length; i++) {
+          highCardExcludedRows[i] = dimsArray[i];
+        }
+
+        outputRow[outputRow.length - 1] =
```

Comment:
What is the data arrangement inside outputRow? It seems like first putting measure columns, then no dictionary columns and last element is dictionary keys in byte array? Can you comment this outputRow format in the function header